File: src/main/java/io/zeitwert/fm/collaboration/model/impl/ObjNotePersistenceProvider.java

package io.zeitwert.fm.collaboration.model.impl;

import static io.zeitwert.ddd.util.Check.requireThis;

import org.jooq.DSLContext;
import org.jooq.exception.NoDataFoundException;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.DependsOn;

import io.zeitwert.ddd.persistence.jooq.AggregateState;
import io.zeitwert.ddd.persistence.jooq.base.ObjExtnPersistenceProviderBase;
import io.zeitwert.fm.collaboration.model.ObjNote;
import io.zeitwert.fm.collaboration.model.db.Tables;
import io.zeitwert.fm.collaboration.model.db.tables.records.ObjNoteRecord;

@Configuration("notePersistenceProvider")
@DependsOn("codePartListTypeEnum")
public class ObjNotePersistenceProvider extends ObjExtnPersistenceProviderBase<ObjNote> {

	public ObjNotePersistenceProvider(DSLContext dslContext) {
		super(ObjNote.class, dslContext);
		this.mapField("relatedToId", AggregateState.EXTN, "related_to_id", Integer.class);
		this.mapField("noteType", AggregateState.EXTN, "note_type_id", String.class);
		this.mapField("subject", AggregateState.EXTN, "subject", String.class);
		this.mapField("content", AggregateState.EXTN, "content", String.class);
		this.mapField("isPrivate", AggregateState.EXTN, "is_private", Boolean.class);
	}

	@Override
	protected boolean hasAccount() {
		return false;
	}

	@Override
	public ObjNote doCreate() {
		return this.doCreate(this.dslContext().newRecord(Tables.OBJ_NOTE));
	}

	@Override
	public ObjNote doLoad(Integer objId) {
		requireThis(objId != null, "objId not null");
		ObjNoteRecord noteRecord = this.dslContext().fetchOne(Tables.OBJ_NOTE,
				Tables.OBJ_NOTE.OBJ_ID.eq(objId));
		if (noteRecord == null) {
			throw new NoDataFoundException(this.getClass().getSimpleName() + "[" + objId + "]");
		}
		return this.doLoad(objId, noteRecord);
	}

}

File: fm-domain/src/main/java/io/zeitwert/fm/collaboration/persist/jooq/ObjNotePersistenceProvider.kt
package io.zeitwert.fm.collaboration.persist.jooq

import io.dddrive.core.enums.model.Enumerated
import io.dddrive.core.property.model.BaseProperty
import io.dddrive.core.property.model.EnumProperty
import io.zeitwert.dddrive.ddd.persist.jooq.JooqObjPersistenceProviderBase
import io.zeitwert.fm.collaboration.model.ObjNote
import io.zeitwert.fm.collaboration.model.ObjNoteRepository
import io.zeitwert.fm.collaboration.model.db.Tables
import io.zeitwert.fm.collaboration.model.enums.CodeNoteType
import io.zeitwert.fm.collaboration.model.enums.CodeNoteTypeEnum
import io.zeitwert.fm.obj.model.db.Sequences
import org.jooq.DSLContext
import org.jooq.UpdatableRecord
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Lazy
import org.springframework.stereotype.Component

/**
 * jOOQ-based persistence provider for ObjNote aggregates.
 */
@Component("objNotePersistenceProvider")
open class ObjNotePersistenceProvider : JooqObjPersistenceProviderBase<ObjNote>() {

    private lateinit var _dslContext: DSLContext
    private lateinit var _repository: ObjNoteRepository

    @Autowired
    fun setDslContext(dslContext: DSLContext) {
        this._dslContext = dslContext
    }

    @Autowired
    @Lazy
    fun setRepository(repository: ObjNoteRepository) {
        this._repository = repository
    }

    override fun dslContext(): DSLContext = _dslContext

    override fun getRepository(): ObjNoteRepository = _repository

    override fun getAggregateTypeId(): String = AGGREGATE_TYPE_ID

    override fun nextAggregateId(): Any {
        return dslContext()
            .nextval(Sequences.OBJ_ID_SEQ)
            .toInt()
    }

    override fun fromAggregate(aggregate: ObjNote): UpdatableRecord<*> {
        return createObjRecord(aggregate)
    }

    @Suppress("UNCHECKED_CAST")
    override fun storeExtension(aggregate: ObjNote) {
        val objId = aggregate.id as Int

        val existingRecord = dslContext().fetchOne(
            Tables.OBJ_NOTE,
            Tables.OBJ_NOTE.OBJ_ID.eq(objId)
        )

        val record = existingRecord ?: dslContext().newRecord(Tables.OBJ_NOTE)

        record.objId = objId
        record.tenantId = aggregate.tenantId as? Int
        record.relatedToId = (aggregate.getProperty("relatedToId") as? BaseProperty<Int?>)?.value
        record.noteTypeId = (aggregate.getProperty("noteType") as? EnumProperty<CodeNoteType>)?.value?.id
        record.subject = (aggregate.getProperty("subject") as? BaseProperty<String?>)?.value
        record.content = (aggregate.getProperty("content") as? BaseProperty<String?>)?.value
        record.isPrivate = (aggregate.getProperty("isPrivate") as? BaseProperty<Boolean?>)?.value

        if (existingRecord != null) {
            record.update()
        } else {
            record.insert()
        }
    }

    @Suppress("UNCHECKED_CAST")
    override fun loadExtension(aggregate: ObjNote, objId: Int?) {
        if (objId == null) return

        val record = dslContext().fetchOne(
            Tables.OBJ_NOTE,
            Tables.OBJ_NOTE.OBJ_ID.eq(objId)
        ) ?: return

        (aggregate.getProperty("relatedToId") as? BaseProperty<Int?>)?.value = record.relatedToId

        record.noteTypeId?.let { noteTypeId ->
            (aggregate.getProperty("noteType") as? EnumProperty<CodeNoteType>)?.value =
                CodeNoteTypeEnum.getNoteType(noteTypeId)
        }

        (aggregate.getProperty("subject") as? BaseProperty<String?>)?.value = record.subject
        (aggregate.getProperty("content") as? BaseProperty<String?>)?.value = record.content
        (aggregate.getProperty("isPrivate") as? BaseProperty<Boolean?>)?.value = record.isPrivate
    }

    companion object {
        private const val AGGREGATE_TYPE_ID = "obj_note"
    }
}

